feat(cluster connect): add service-registry

rkpattnaik780 opened the pull request:
### Description

The PR adds the following features:
1. Change code structure to allow cluster commands to work with multiple services
2. Add Service registry to cluster connect command
3. Disable `cluster connect` from using the context.

Closes #903 

Next step: follow this pattern for `cluster bind`

### Verification Steps
<!-- Add verification steps here if applicable. Remove this section if it does not apply -->

1. Login to your cluster using `oc login`.
2. Run cluster connect command to connect a Kafka instance
```
./rhoas cluster connect --service-type kafka --service-id c58p7j6snr2tj7aleaf0
```
3. Run cluster connect command to connect a service registry instance
```
./rhoas cluster connect --service-type service-registry --service-id a6cae630-104d-4f30-b962-a37661b240c4
```

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation change
- [ ] Other (please specify)

### Package structure diagram

![godepgraph_i](https://user-images.githubusercontent.com/23582438/135286553-dd02b583-46d6-4432-963b-bafa9da33ecf.png)


### Checklist

- [X] Documentation added for the feature
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer

wtrocki commented:
What happened here. 
This needs rebase?

wtrocki commented on pkg/cluster/common.go:
```diff
@@ -0,0 +1,95 @@
+	"k8s.io/client-go/kubernetes"
+)
+
+func makeRequest(ctx context.Context, clientSet *kubernetes.Clientset, path string, serviceName string, localizer localize.Localizer) error {
```
makeGetRequest

wtrocki commented on pkg/cluster/common.go:
```diff
@@ -0,0 +1,95 @@
+	return nil
+}
+
+func createResourceRequest(ctx context.Context, clientSet *kubernetes.Clientset, path string, serviceName string, crJSON []byte) error {
```
Same postKubernetesRequest

wtrocki commented on pkg/cluster/common.go:
```diff
@@ -0,0 +1,95 @@
+	return nil
+}
+
+func watchGeneric(w watch.Interface, namespace string, crName string, opts Options) error {
```
To generic name :)

wtrocki commented on pkg/cluster/service.go:
```diff
@@ -0,0 +1,129 @@
+	CreateResource(ctx context.Context, c *KubernetesCluster, namespace string, serviceID string, opts Options) error
+}
+
+type Kafka struct{}
```
@craicoverflow is expert for golang patterns, but.. I think purpose of declaring interface is that it can be used as generic dependency - implementations should be in separate files.

Mixing kafka and service registry in single cluster package breaks our main rule - ability to separate services and fast addition of new ones.

wtrocki commented on pkg/cluster/service.go:
```diff
@@ -0,0 +1,129 @@
+
+}
+
+type ServiceRegistry struct{}
```
ServiceRegistryResource?

rkpattnaik780 replied:
Using Connection suffix instead.

wtrocki commented on pkg/cluster/service.go:
```diff
@@ -0,0 +1,129 @@
+
+	path := kafka.GetKafkaConnectionsAPIURL(namespace)
+
+	kafkaInstance, _, err := api.Kafka().GetKafkaById(ctx, serviceID).Execute()
```
My take is that we should validate service id before - once we creating resource we should not print out such breaking validation errors.

wtrocki commented on pkg/cluster/service.go:
```diff
@@ -0,0 +1,129 @@
+	// opts.Logger.Info(opts.Localizer.MustLocalize("cluster.kubernetes.watchForKafkaStatus.log.info.wait"))
+	resource := kafka.AKCResource
+
+	w, err := c.dynamicClient.Resource(resource).Namespace(namespace).Watch(context.TODO(), metav1.ListOptions{
```
this statement can be moved to generic method - I do not see anything specific to service here.

wtrocki commented:
Very good quality contribution. 

✅ Style
✅ Abstraction
✅ Simplicity
✅ Future proof
🚫 Package structure/separation between services

wtrocki commented:
Will verify after rebase :D

wtrocki commented:
Let's resolve the comments that were addressed. Provide proper verification steps and remove WIP from PR.

wtrocki reviewed (changes requested):
I do not see service binding integration. We need to update service binding command to support both resources. We cannot merge this right now as it will not enable devs to complete their experience.

wtrocki commented:
constants package name doesn't well reflect what that package does. There is more into that package

wtrocki commented on pkg/cluster/constants/serviceregistry/ServiceRegistryConnection.go:
```diff
@@ -0,0 +1,51 @@
+}
+
+// BootstrapServerSpec contains server host information that can be used to connecto the  Kafka
+type BootstrapServerSpec struct {
```
This is not used by registry and should be removed.

wtrocki commented on pkg/cluster/constants/serviceregistry/ServiceRegistryConnection.go:
```diff
@@ -0,0 +1,51 @@
+}
+
+// ServiceRegsitryConnectionList contains a list of ServiceRegsitryConnection
+type ServiceRegsitryConnectionList struct {
```
I do not see List being used. (It might be) worth remove list from Kafka and registry

wtrocki commented on pkg/cluster/cluster.go:
```diff
@@ -8,13 +8,14 @@ type ConnectArguments struct {
 	OfflineAccessToken      string
```
Would you be able to draw package structure diagram from the tool and attach it to PR?

wtrocki commented on pkg/cluster/kafkaservice/kafkaservice.go:
```diff
@@ -0,0 +1,65 @@
+	"github.com/redhat-developer/app-services-cli/pkg/kafka/kafkaerr"
+)
+
+type KafkaConnection struct {
```
Missing comments. Also this will clash with actual Connection object in constants. Let's try to avoid using the same names across the board. Specially that this is not DTO/Entity but functional/service class

wtrocki commented on pkg/cluster/constants/kafka/kcManager.go:
```diff
@@ -0,0 +1,52 @@
+	Resource: "kafkaconnections",
+}
+
+var tokenSecretName = "rh-cloud-services-accesstoken-cli"
```
[ResponsibilityLeak]
This should not be part of the class. 
Token secret is happening before service is picked up and it should be shared

wtrocki commented on pkg/cluster/constants/kafka/kcManager.go:
```diff
@@ -0,0 +1,52 @@
+var tokenSecretName = "rh-cloud-services-accesstoken-cli"
+
+/*  #nosec */
+var serviceAccountSecretName = "rh-cloud-services-service-account"
```
Should not be shared

wtrocki commented on pkg/cluster/constants/kafka/kcManager.go:
```diff
@@ -0,0 +1,52 @@
+	return fmt.Sprintf("/apis/rhoas.redhat.com/v1alpha1/namespaces/%v/kafkaconnections", namespace)
+}
+
+func CreateKCObject(crName string, namespace string, kafkaID string) *KafkaConnection {
```
I question value of this being in constants. 
It can be part of the service - is that used for some circular dependency avoidance?

wtrocki commented on pkg/cluster/kafkaservice/kafkaservice.go:
```diff
@@ -0,0 +1,65 @@
+	return err
+}
+
+func GetWatchErrorMessages() map[string]string {
```
[Question]

What is purpose of that function - do we need function for something that should be constant?
Is there any benefit for having slice vs explicit error strings (having this this way will affect our linter capability)
Best to avoid it unless there is reason for it

wtrocki replied:
Oh.. this is not even used in this file, which is also broken responsibility. Keys should be defined as close to the actual usage.

wtrocki commented on pkg/cluster/registryservice/registryservice.go:
```diff
@@ -0,0 +1,64 @@
+	return err
+}
+
+func getWatchErrorMessages() map[string]string {
```
I think I suspect that we wanted some way to control errors happening for the same service. 

This requires dicussion - it might be valid code but it looks hacky and strange without context

wtrocki commented on pkg/cluster/utils/utils.go:
```diff
@@ -0,0 +1,89 @@
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+func ResourceExists(ctx context.Context, c *cluster.KubernetesCluster, path string, serviceName string, opts cluster.Options) error {
```
Utils should be used across the board and form SDK. Comments are strongly required to understand what this does without checking code.

craicoverflow replied:
What is `Resource` here? This function needs to be documented.

craicoverflow replied:
This returns an error but it should return a bool, or a bool + error.

rkpattnaik780 replied:
Sticking to err as adding a bool seems bit redundant.

wtrocki commented on pkg/cluster/cluster.go:
```diff
@@ -8,13 +8,14 @@ type ConnectArguments struct {
 	IgnoreContext           bool
-	SelectedKafka           string
 	Namespace               string
+	SelectedService         string
```
[Responsibility]
Selected service is not argument/input (it is actually part of the logic). If someone uses this as lib it is not clear which one should be used (ID) and if SelectedService will be taken into account

rkpattnaik780 replied:
Can you elaborate this a bit?

rkpattnaik780 replied:
Closing this discussion as separation entails additional function argument.

wtrocki commented on pkg/cluster/connectionsManager.go:
```diff
@@ -0,0 +1,68 @@
+)
+
+// IsKCInstalledOnCluster checks the cluster to see if a KafkaConnection CRD is installed
+func IsKCInstalledOnCluster(ctx context.Context, c *KubernetesCluster) (bool, error) {
```
function probably should be renamed to 
`IsResourceInstalledonCluster` and accept resource from constants and then possibly moved to utils?

rkpattnaik780 replied:
Noticed we would need to check for SBO as well.

wtrocki replied:
Yes! :D

rkpattnaik780 replied:
Facing issues clubbing them to utils. But this is indeed no place to be keeping them.

wtrocki commented on pkg/cluster/custom_resource.go:
```diff
@@ -0,0 +1,8 @@
+package cluster
```
Very strainge file name

wtrocki commented on pkg/cluster/custom_resource.go:
```diff
@@ -0,0 +1,8 @@
+
+import "context"
+
+type CustomConnection interface {
```
Every key component like interface needs documentation

wtrocki commented on pkg/cluster/kubernetes_cluster.go:
```diff
@@ -118,19 +112,12 @@ func (c *KubernetesCluster) CurrentNamespace() (string, error) {
 	return namespace, err
 }
 
+// nolint:funlen
```
:) Strangely enough function is short so seems like leftover.

wtrocki commented on pkg/cluster/kubernetes_cluster.go:
```diff
@@ -35,17 +32,19 @@ import (
+	DynamicClient      dynamic.Interface
 	kubeconfigLocation string
 }
 
```
Bellow we have duplicated constants:
tokenSecretName etc.

wtrocki commented on pkg/cmd/cluster/connect/connect.go:
```diff
@@ -48,9 +53,6 @@ func NewConnectCommand(f *factory.Factory) *cobra.Command {
 		Example: opts.localizer.MustLocalize("cluster.connect.cmd.example"),
 		Args:    cobra.NoArgs,
 		RunE: func(cmd *cobra.Command, _ []string) error {
-			if opts.ignoreContext == true && !opts.IO.CanPrompt() {
```
Note - this requires documentation change and better documentation in the cluster root cmd

wtrocki commented on pkg/cmd/cluster/connect/connect.go:
```diff
@@ -59,7 +61,11 @@ func NewConnectCommand(f *factory.Factory) *cobra.Command {
+	cmd.Flags().StringVar(&opts.serviceID, "service-id", "", opts.localizer.MustLocalize("cluster.common.flag.serviceId.description"))
+	cmd.Flags().StringVar(&opts.serviceType, "service-type", "", opts.localizer.MustLocalize("cluster.common.flag.serviceType.description"))
+
+	_ = cmd.MarkFlagRequired("service-type")
```
How this works with context mode where user is asked about what service they want?

We need that for backwards compatibility etc.

rkpattnaik780 replied:
Context mode to be restored in a subsequent PR.

wtrocki commented on pkg/cmd/cluster/connect/connect.go:
```diff
@@ -70,43 +76,64 @@ func runConnect(opts *options) error {
+	case "kafka":
+		_, _, err := kafka.GetKafkaByID(ctx, api.Kafka(), serviceID)
+		if err != nil {
+			return err
```
That is going to send very unclear code to user.

wtrocki commented on pkg/cmd/cluster/connect/connect.go:
```diff
@@ -70,43 +76,64 @@ func runConnect(opts *options) error {
 	return nil
 }
+
+func validateServiceID(ctx context.Context, api *api.API, serviceType string, serviceID string) error {
```
Can validation be part of the connect

wtrocki commented on pkg/cmd/cluster/connect/connect.go:
```diff
@@ -70,43 +76,64 @@ func runConnect(opts *options) error {
+
+	var service cluster.CustomConnection
+
+	switch opts.serviceType {
```
[ConstantsLeak]

There is multiple places where we iterate over Kafka/Registry types. It will be better if we do it once and create instances of those objects. 
Then instances can be used for validation etc.

This is leaking logic out of the abstraction

wtrocki replied:
Every time we iterate over types 💡  should appear with `Should We use Strategy pattern here`?

wtrocki commented:
We are missing:

- Documentation
- Context mode
- Service binding integration

craicoverflow commented on pkg/cluster/cluster.go:
```diff
@@ -8,13 +8,14 @@ type ConnectArguments struct {
 // Cluster defines methods used to interact with a cluster
 type Cluster interface {
-	Connect(ctx context.Context, opts *ConnectArguments) error
+	Connect(ctx context.Context, connectOpts *ConnectArguments, c CustomConnection, opts Options) error
```
I have not dived into the code in my editor to see what's happening here, but why is there a need for so many arguments? And then two sets of options?

We should be striving for 2 arguments maximum. There are going to be exceptions to that, but does this have to have this many?

wtrocki replied:
Context is not needed - should be in options. We have two sets of arguments because with Rama we try to keep structures with responsibilities separate. Input arguments and kubernetes client that can be reused across different commands etc.

craicoverflow commented on pkg/cluster/connectionsManager.go:
```diff
@@ -0,0 +1,68 @@
+	return true, data.Error()
+}
+
+func CheckIfConnectionExists(ctx context.Context, c *KubernetesCluster, namespace string, service interface{}, serviceName string, opts Options) error {
```
```go
func CheckIfConnectionExists(ctx context.Context, c *KubernetesCluster, namespace string, service ManagedService, serviceName string, opts Options) error {
```

I think there is space/opportunity for creating an actual interface here. We should not really use `interface{}` in most cases.

craicoverflow replied:
Too many arguments here.

`CheckIfConnectionExists` could become a member of the `Service` interface I suggested.

```go
if service.ConnectionExists() {
	// do something
}
```

craicoverflow commented on pkg/cluster/connectionsManager.go:
```diff
@@ -0,0 +1,68 @@
+	var path string
+	var status int
+
+	switch service.(type) {
+	case *srsmgmtv1.Registry:
+		path = serviceregistry.GetServiceRegistryAPIURL(namespace)
+	case kafkamgmtclient.KafkaRequest:
+		path = kafka.GetKafkaConnectionsAPIURL(namespace)
+	}
```
```go
type Service interface {
  ApiURL() string
  Status() ServiceStatus
}
```

We could have a service interface, and create concrete types for each implementation like `KafkaService`, `RegistryService`.

craicoverflow commented on pkg/cluster/constants/serviceregistry/ServiceRegistryConnection.go:
```diff
@@ -0,0 +1,51 @@
+}
+
+// ServiceRegsitryConnectionStatus defines the observed state of ServiceRegsitryConnection
+type ServiceRegsitryConnectionStatus struct {
```
```suggestion
type ServiceRegistryConnectionStatus struct {
```
typo

craicoverflow commented on pkg/cluster/constants/serviceregistry/ServiceRegistryConnection.go:
```diff
@@ -0,0 +1,51 @@
+}
+
+// ServiceRegsitryConnection schema
+type ServiceRegsitryConnection struct {
```
typo

craicoverflow commented on pkg/cluster/constants/serviceregistry/ServiceRegistryConnection.go:
```diff
@@ -0,0 +1,51 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ServiceRegsitryConnectionSpec   `json:"spec,omitempty"`
+	Status ServiceRegsitryConnectionStatus `json:"status,omitempty"`
```
typo

craicoverflow commented on pkg/cluster/constants/serviceregistry/srManager.go:
```diff
@@ -0,0 +1,52 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var (
+	SRCGroup   = "rhoas.redhat.com"
+	SRCVersion = "v1alpha1"
+)
+
+var RegistryResourceMeta = metav1.TypeMeta{
+	Kind:       "ServiceRegistryConnection",
+	APIVersion: SRCGroup + "/" + SRCVersion,
+}
+
+var tokenSecretName = "rh-cloud-services-accesstoken-cli"
+
+/*  #nosec */
+var serviceAccountSecretName = "rh-cloud-services-service-account"
+
+var SRCResource = schema.GroupVersionResource{
+	Group:    SRCGroup,
+	Version:  SRCVersion,
+	Resource: "serviceregistryconnections",
+}
+
+func GetServiceRegistryAPIURL(namespace string) string {
+	return fmt.Sprintf("/apis/rhoas.redhat.com/v1alpha1/namespaces/%v/serviceregistryconnections", namespace)
+}
```
None of these are constants.

craicoverflow commented on pkg/cluster/kafkaservice/kafkaservice.go:
```diff
@@ -0,0 +1,65 @@
+	return err
+}
+
+func (k *KafkaConnection) CreateCustomResource(ctx context.Context, c *cluster.KubernetesCluster, serviceID string, namespace string, opts cluster.Options) error {
```
I feel that an options object is not needed here. It is very implementation specific. For example, what if I want to create a `XConnection` and not use `Localizer`?

My vote is to remove use of `options` from any interface methods, they can be passes as properties of the concrete type (KafkaConnection).

wtrocki replied:
Very very good point that follows good patterns!

craicoverflow commented on pkg/cluster/kubernetes_cluster.go:
```diff
@@ -97,14 +96,9 @@ func NewKubernetesClusterConnection(connection connection.Connection,
 		dynamicClient,
-		io,
-		localizer,
 		kubeconfig,
```
I don't understand why connection etc are being removed in favour of being passed as function arguments?

wtrocki commented:
@rkpattnaik780 I accidentially overriden your changes. Please force push your local branch.

wtrocki commented:
lets close addressed comments and leave extra comments if we do not address it.
Unfortunate that we killed cluster as this could have been verified and merged today :)

rkpattnaik780 commented:
> lets close addressed comments and leave extra comments if we do not address it. Unfortunate that we killed cluster as this could have been verified and merged today :)

Still needs some refactorings, the bind commands have a scope of improvements and context needs to be restored.

wtrocki commented:
Let's work on this items first and prepare entire PR for final review. I would like to avoid numerous reviews as it takes 1 hour to review and verify changes.

rkpattnaik780 commented on pkg/cluster/connectionsManager.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+// IsSBOInstalledOnCluster checks the cluster to see if ServiceBinding CRD is installed
+func IsSBOInstalledOnCluster(ctx context.Context, c *KubernetesCluster) (bool, error) {
```
@wtrocki have added methods here to check for Kafka CRD, Registry CRD and SBO installation. Moving in separate packages was causing issues.

rkpattnaik780 commented on pkg/cmd/cluster/bind/bind.go:
```diff
@@ -69,47 +71,41 @@ func NewBindCommand(f *factory.Factory) *cobra.Command {
 
-	err = cluster.ExecuteServiceBinding(opts.Context, opts.Logger, opts.localizer, &cluster.ServiceBindingOptions{
-		ServiceName:             kafkaInstance.GetName(),
+	err = cluster.ExecuteServiceBinding(opts.Context, opts.Logger, opts.localizer, service, &cluster.ServiceBindingOptions{
```
Unlike connect method, `ExecuteServiceBinding` is not a part of cluster struct, is there a reason for this?

wtrocki replied:
Historical. Separation of concerns but not valid anymore

rkpattnaik780 commented on pkg/cluster/custom_resource.go:
```diff
@@ -0,0 +1,14 @@
+	CreateCustomResource(ctx context.Context, c *KubernetesCluster, serviceID string) error
+	CustomResourceExists(ctx context.Context, c *KubernetesCluster, serviceID string) error
+	BindCustomConnection(ctx context.Context, serviceName string, options ServiceBindingOptions, clients *KubernetesClients) error
+	CustomConnectionExists(ctx context.Context, dynamicClient dynamic.Interface, serviceName string, ns string) error
```
It seems CustomConnectionExists and CustomResourceExists tend to do the same thing in different way, will it make sense to have one method for it?

wtrocki replied:
yes

wtrocki commented:
Will do separate PR
